### ISSUE TITLE
music 29 lab 8: frequency domain model builder

### ISSUE THREAD

**Issue by EdmundCampion** [CNMAT/CNMAT-MMJ-Depot#55]
This is assignment to build a Spectral Modeling patch/app for Music 29 SP2015.  The due date is no later than February 20th, 2015.

Here are basic details, but please sit down with Ed to go over everything before starting.

Description: 

Load a soundfile of a single instrumental sample playing one note  (e.g. a Sather Bell strike or a Saxophone playing G etc..).  Playback of the sample should be on separate gain, have loop/no loop, have mute/unmute button.

Prefader gain from sample is sent to sonogram analysis onboard with patch.

Using sonogram analysis and probing sinusoid locate specific frequencies inside the sample.  (This will entail going back to frequency domain analysis patch in MUSIC 29 and working out some better and more refined way to capture probed frequencies)

Once a frequency is located it gets added to an individual frequency control track that can post the frequency selected, play the frequency constantly, have full gain control, and have a mute/unmute button.  Output should be mono.

Once many frequency are located, all the frequencies should appear as a  mixing board, allowing for fine tuning of amplitudes and control of every frequency found. (maybe an upper limit of 20 frequencies?).

The mixed output of all frequencies and the original source sample should be sent to a dual master mono output, one for sinusoids, one for sample.   The normal default of the master mono outpust gain is a slider output for each.  The sinusoid master output will have a button that selects an ADSR feature.   When the ADSR component is checked, then a window is enable to create a specific envelope that will auto match to the length of the currently loaded soundfile and will trigger each time the loaded soundfile begins.

**Comment by equilet** [comment]
note: consider freq domain analysis as a plunder-space for this.

**Comment by equilet** [comment]
NEW DEADLINE: Feb 8th so that Matt can test.

**Comment by equilet** [comment]
OK, this is in the depot now, and included in the overview.  @mattschumaker @EdmundCampion can you please test and let me know how you find the functionality.  I'm certain that it will need labels, etc.  

Don't we all.  :)

**Comment by mattschumaker** [comment]
Hi Jeff,

This is what I am getting so far.  I had two crashes on the standalone.  Pls. see below.  

Program looks beautiful.
Loading and playing back sounds works. Playback looping works.
Directions are clear.

While following directions, standalone program crashed on me twice. Before crashing, gave me these errors

Error:
playlist~: couldn't find /Users/jml/Documents/CNMAT/repositories/CNMAT-Production/campion/natural_frequencies/lib/audio/SatherBells2014_Samples-ord/2_02bell.aif

errors when trying to press range area:
"address is unbound: length: address /markers is unbound
error evaluating expression: "

errors when moving a partial id:
address is unbound: scale: address /tempmarks is unbound
error evaluating expression:


Recommend labels (as you suspected…however with a couple of labels would be very functional for class next week bk gsis can explain.)
label for range area
label for partial ids & sliders
possible labels for sound file volume and model volumes--but these are explained in the help file.

**Comment by equilet** [comment]
Please provide a list of steps to reproduce the crash.  Thanks!

**Comment by mattschumaker** [comment]
Started app again.  These steps led me to a crash:

--Open application

--see error for missing sound on max console: 
playlist~: couldn't find /Users/jml/Documents/CNMAT/repositories/CNMAT-Production/campion/natural_frequencies/lib/audio/SatherBells2014_Samples-ord/2_02bell.aif

--open subpatcher quickinfo

--drag sound file z14 bell to playlist area.  (This is an .aiff file)

--turn on audio

--turn up sound file volume slider

--hit loop sound.  

--working on following further directions when the app crashed.  The short sound looped approximately five times. In the past, I had been able to do more, like select a range button on the right, before app crashes.

**Comment by equilet** [comment]
I can't get this to crash.  Will follow up with you in person.

**Comment by equilet** [comment]
Matt and I tested this, got the crash to happen.  I isolated it down to either a crash on cycling's part, or a memleak that odot exhibited a while ago.  Either way, the crash doesn't occur any more...  I'll follow up for my own edification, but for now this is sorted out and the model builder will be deployed to the lab.  Rev 1 is finished and annotated with badge/banner/comments/labels.  Rev 2 or any further work will be another issue.

**Comment by EdmundCampion** [comment]
Looking over the Freq Model builder standalone.
Thanks Jeff for some very nice improvements in GUI on that.
There are a few issues to deal with.

Please change the name of the 29 standalone and Depot patch to:  Freq_Model_Builder 

•  always include a link to the DSP Settings and always include a background default message for those settings.

•  For the "audio button"  make sure there is some evidence that the audio is running as we have been doing.

Both of these problems above were solved with the Audio I/O.  Right now, the audio button is initiated as "green" which to me means "on".  

•  if possible implement a way to jump the start point of the audio playback.
•  implement a "pause" "resume" soundfile playback with keyboard space bar
•  if possible implement a loop feature so that a particular section can be selected and looped.
•  As mentioned before, we will need to add <frequency in Hz> <MIDI cents><amplitude> for each of the sinusoids. That should be posted right next to the sinusoid numbering that is so nicely implemented in the tool.  
•  add a "write" to save out the data to text file.  See if you can make it so that you have two lists  -- one with <MIDI cents> <freq> <amp>  and a second that is formatted for direct input into sinusoids~ 

•  if possible, the resizing of the sonogram should also reposition the sinusoid freq/gain pots to the right position.

•  As mentioned, we need the "nudge" tool for finegrain adjustment of sinusoids

•  please make sure that stereo files are being summed and not just playing L or R.

**Comment by mattschumaker** [comment]
This Lab 7 standalone is in the Wheeler lab.  Basic functionality seems to be working great on the instructor and student machines. 

Introduced Gsis to this lab in meeting today.  Gsis very impressed and excited by this new standalone!

Rafael pointed out one repeatable bug which will not affect labs this week:
In the soundfile window, if you press "clear" while the sound is looping then standalone will crash. Easy to avoid this for the moment.

Expecting lab to go well.  Looking forward to watching students encounter this lab later this week.  We should get good feedback from their use.

**Comment by equilet** [comment]
background dsp initialization, nudging, define start point, and loop are all implemented.

- click on the recycling icon for selection and playback of a loop
- make a selection to define playback bounds
- type the right and left arrow keyboard keys with a marker selected and gain up for auditioning - the frequency is nudged by a "nudge unit".  if you'd like for me to make this nudge value finer grain, i can do that.
- m29dspinit is now a patch that will be included in all m29 software
